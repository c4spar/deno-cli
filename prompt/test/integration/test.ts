import {
  assert,
  assertSnapshot,
  copy,
  dirname,
  expandGlob,
  lt,
  WalkEntry,
} from "../../../dev_deps.ts";

const baseDir = `${dirname(import.meta.url).replace("file://", "")}`;

for await (const file: WalkEntry of expandGlob(`${baseDir}/fixtures/*.ts`)) {
  if (file.isFile) {
    const name = file.name.replace(/_/g, " ").replace(".ts", "");
    Deno.test({
      name: `prompt - integration - ${name}`,
      ignore: lt(Deno.version.deno, "1.10.0"),
      async fn(t) {
        const output: string = await runPrompt(file);
        const os = Deno.build.os === "windows" ? ".windows" : "";
        await assertSnapshot(t, output, {
          path: `__snapshots__/test.ts${os}.snap`,
        });
      },
    });
  }
}

function getCmdFlagsForFile(file: WalkEntry): string[] {
  if (file.name === "input_no_location_flag.ts") {
    return [
      "--unstable",
      "--allow-all",
    ];
  }
  return [
    "--unstable",
    "--allow-all",
    "--location",
    "https://cliffy.io",
  ];
}

async function runPrompt(file: WalkEntry): Promise<string> {
  const inputPath: string = file.path.replace(/\.ts$/, ".in");
  const inputFile = await Deno.open(inputPath);
  const flags = getCmdFlagsForFile(file);
  const process = Deno.run({
    stdin: "piped",
    stdout: "piped",
    cmd: [
      "deno",
      "run",
      ...flags,
      file.path,
    ],
    env: {
      NO_COLOR: "true",
    },
  });

  const [output, bytesCopied] = await Promise.all([
    process.output(),
    copy(inputFile, process.stdin),
  ]);
  inputFile.close();
  process.stdin.close();
  process.close();

  assert(bytesCopied > 0, "No bytes copied");

  return new TextDecoder().decode(output);
}
