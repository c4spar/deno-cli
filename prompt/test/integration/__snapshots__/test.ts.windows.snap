export const snapshot = {};

snapshot[`prompt - integration - checkbox 1`] = `
"? Select an option
❯ × Foo
  × Bar
  × Baz
\\x1b[1A\\x1b[3A\\x1b[19G\\x1b[?25l\\x1b[G\\x1b[0J
? Select an option
  × Foo
❯ × Bar
  × Baz
\\x1b[1A\\x1b[3A\\x1b[19G\\x1b[?25l\\x1b[G\\x1b[0J
? Select an option
  × Foo
  × Bar
❯ × Baz
\\x1b[1A\\x1b[3A\\x1b[19G\\x1b[?25l\\x1b[G\\x1b[0J
? Select an option
  × Foo
  × Bar
❯ √ Baz
\\x1b[1A\\x1b[3A\\x1b[19G\\x1b[?25l\\x1b[G\\x1b[0J
? Select an option » Baz
\\x1b[?25h\\x1b[?25h"
`;

snapshot[`prompt - integration - confirm 1`] = `
"? Please confirm (y/N) » 
❯ some hint
\\x1b[1A\\x1b[1A\\x1b[26G\\x1b[G\\x1b[0J
? Please confirm (y/N) » yes
❯ some hint
\\x1b[1A\\x1b[1A\\x1b[27G\\x1b[G\\x1b[0J
? Please confirm (y/N) » Yes
\\x1b[?25h\\x1b[?25h"
`;

snapshot[`prompt - integration - confirm 2`] = `
"? Please confirm (y/N) » 
❯ some hint
\\x1b[1A\\x1b[1A\\x1b[26G\\x1b[G\\x1b[0J
? Please confirm (y/N) » no
❯ some hint
\\x1b[1A\\x1b[1A\\x1b[27G\\x1b[G\\x1b[0J
? Please confirm (y/N) » No
\\x1b[?25h\\x1b[?25h"
`;

snapshot[`prompt - integration - confirm 3`] = `
"? Please confirm (y/N) » 
❯ some hint
\\x1b[1A\\x1b[1A\\x1b[26G\\x1b[G\\x1b[0J
? Please confirm (y/N) » No
\\x1b[?25h\\x1b[?25h"
`;

snapshot[`prompt - integration - input 1`] = `
"? Whats your name? (foo) » 
\\x1b[1A\\x1b[28G\\x1b[G\\x1b[0J
? Whats your name? (foo) » foo bar
\\x1b[1A\\x1b[35G\\x1b[G\\x1b[0J
? Whats your name? (foo) » foo bar
\\x1b[?25h\\x1b[?25h"
`;

snapshot[`prompt - integration - input no location flag 1`] = `
"? Works without --location? (hope so) » 
\\x1b[1A\\x1b[41G\\x1b[G\\x1b[0J
? Works without --location? (hope so) » yes
\\x1b[1A\\x1b[44G\\x1b[G\\x1b[0J
? Works without --location? (hope so) » yes
\\x1b[?25h\\x1b[?25h"
`;

snapshot[`prompt - integration - input no prefix 1`] = `
"Whats your name? (foo) » 
\\x1b[1A\\x1b[26G\\x1b[G\\x1b[0J
Whats your name? (foo) » bar
\\x1b[1A\\x1b[29G\\x1b[G\\x1b[0J
Whats your name? (foo) » bar
\\x1b[?25h\\x1b[?25h"
`;

snapshot[`prompt - integration - input prefix 1`] = `
"PREFIX Whats your name? (foo) » 
\\x1b[1A\\x1b[33G\\x1b[G\\x1b[0J
PREFIX Whats your name? (foo) » bar
\\x1b[1A\\x1b[36G\\x1b[G\\x1b[0J
PREFIX Whats your name? (foo) » bar
\\x1b[?25h\\x1b[?25h"
`;

snapshot[`prompt - integration - input suggestions 1`] = `
"? Whats your name? (foo) » 
  foo
  bar
  baz
\\x1b[1A\\x1b[3A\\x1b[28G\\x1b[G\\x1b[0J
? Whats your name? (foo) » foo
❯ foo
\\x1b[1A\\x1b[1A\\x1b[31G\\x1b[G\\x1b[0J
? Whats your name? (foo) » foo
\\x1b[?25h\\x1b[?25h"
`;

snapshot[`prompt - integration - list 1`] = `
"? Enter some keywords (foo, far) » 
❯ some hint...
\\x1b[1A\\x1b[1A\\x1b[36G\\x1b[G\\x1b[0J
? Enter some keywords (foo, far) » foo, bar
❯ some hint...
\\x1b[1A\\x1b[1A\\x1b[44G\\x1b[G\\x1b[0J
? Enter some keywords (foo, far) » foo, bar
\\x1b[?25h\\x1b[?25h"
`;

snapshot[`prompt - integration - number 1`] = `
"? How old are you? (7) » 
\\x1b[1A\\x1b[26G\\x1b[G\\x1b[0J
? How old are you? (7) » 19
\\x1b[1A\\x1b[28G\\x1b[G\\x1b[0J
? How old are you? (7) » 20
\\x1b[1A\\x1b[27G\\x1b[G\\x1b[0J
? How old are you? (7) » 21
\\x1b[1A\\x1b[27G\\x1b[G\\x1b[0J
? How old are you? (7) » 20
\\x1b[1A\\x1b[27G\\x1b[G\\x1b[0J
? How old are you? (7) » 19
\\x1b[1A\\x1b[27G\\x1b[G\\x1b[0J
? How old are you? (7) » 18
\\x1b[1A\\x1b[27G\\x1b[G\\x1b[0J
? How old are you? (7) » 17
\\x1b[1A\\x1b[27G\\x1b[G\\x1b[0J
? How old are you? (7) » 18
\\x1b[1A\\x1b[27G\\x1b[G\\x1b[0J
? How old are you? (7) » 19
\\x1b[1A\\x1b[27G\\x1b[G\\x1b[0J
? How old are you? (7) » 20
\\x1b[1A\\x1b[27G\\x1b[G\\x1b[0J
? How old are you? (7) » 20
\\x1b[?25h\\x1b[?25h"
`;

snapshot[`prompt - integration - number decimal 1`] = `
"? How old are you? (7) » 
\\x1b[1A\\x1b[26G\\x1b[G\\x1b[0J
? How old are you? (7) » 19.
\\x1b[1A\\x1b[29G\\x1b[G\\x1b[0J
? How old are you? (7) » 19.1
\\x1b[1A\\x1b[29G\\x1b[G\\x1b[0J
? How old are you? (7) » 19.2
\\x1b[1A\\x1b[29G\\x1b[G\\x1b[0J
? How old are you? (7) » 19.3
\\x1b[1A\\x1b[29G\\x1b[G\\x1b[0J
? How old are you? (7) » 19.4
\\x1b[1A\\x1b[29G\\x1b[G\\x1b[0J
? How old are you? (7) » 19.4
\\x1b[?25h\\x1b[?25h"
`;

snapshot[`prompt - integration - secret 1`] = `
"? enter your secret » 
\\x1b[1A\\x1b[23G\\x1b[G\\x1b[0J
? enter your secret » ***
\\x1b[1A\\x1b[26G\\x1b[G\\x1b[0J
? enter your secret » ***
\\x1b[?25h\\x1b[?25h"
`;

snapshot[`prompt - integration - select 1`] = `
"? Select an option
❯ Foo
  Bar
  Baz
\\x1b[1A\\x1b[3A\\x1b[19G\\x1b[?25l\\x1b[G\\x1b[0J
? Select an option
  Foo
❯ Bar
  Baz
\\x1b[1A\\x1b[3A\\x1b[19G\\x1b[?25l\\x1b[G\\x1b[0J
? Select an option
  Foo
  Bar
❯ Baz
\\x1b[1A\\x1b[3A\\x1b[19G\\x1b[?25l\\x1b[G\\x1b[0J
? Select an option » Baz
\\x1b[?25h\\x1b[?25h"
`;

snapshot[`prompt - integration - toggle 1`] = `
"? Please confirm (No) » No / Yes
❯ some hint
\\x1b[1A\\x1b[1A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Please confirm (No) » No / Yes
❯ some hint
\\x1b[1A\\x1b[1A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Please confirm (No) » No / Yes
❯ some hint
\\x1b[1A\\x1b[1A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Please confirm (No) » No / Yes
❯ some hint
\\x1b[1A\\x1b[1A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Please confirm (No) » Yes
\\x1b[?25h\\x1b[?25h"
`;
